File: test-app/src/3algo_backend.py
import math
import itertools

def calculate_distance(city1, city2):
    lat1, lon1 = city1["lat"], city1["lon"]
    lat2, lon2 = city2["lat"], city2["lon"]
    radius = 6371  # Radius of the Earth in kilometers
    dlat = to_radians(lat2 - lat1)
    dlon = to_radians(lon2 - lon1)

    a = math.sin(dlat / 2) ** 2 + math.cos(to_radians(lat1)) * math.cos(to_radians(lat2)) * math.sin(dlon / 2) ** 2
    a = min(1, a)  # Ensure a is within the valid range [0, 1]
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    distance = radius * c

    return distance

def to_radians(degrees):
    return degrees * (math.pi / 180)

def get_distance_matrix(cities):
    num_cities = len(cities)
    distances = [[0] * num_cities for _ in range(num_cities)]

    for i in range(num_cities):
        for j in range(i + 1, num_cities):
            dist = calculate_distance(cities[i], cities[j])
            distances[i][j] = round(dist)
            distances[j][i] = round(dist)  # Ensure symmetry

    return distances

def calculate_tour_distance(tour, distances):
    total_distance = 0
    for i in range(len(tour) - 1):
        from_city = tour[i]
        to_city = tour[i + 1]
        total_distance += distances[from_city][to_city]
    # Complete the tour by returning to the starting city
    total_distance += distances[tour[-1]][tour[0]] if distances[tour[-1]][tour[0]] is not None else 0
    return total_distance

def solve_tsp_nearest(distances):
    num_cities = len(distances)
    visited = [False] * num_cities
    tour = []
    total_distance = 0

    # Start at the first city
    current_city = 0
    tour.append(current_city)
    visited[current_city] = True

    # Repeat until all cities have been visited
    while len(tour) < num_cities:
        nearest_city = None
        nearest_distance = float("inf")

        # Find the nearest unvisited city
        for city in range(num_cities):
            if not visited[city] and distances[current_city] and distances[current_city][city] is not None:
                distance = distances[current_city][city]
                if distance < nearest_distance:
                    nearest_city = city
                    nearest_distance = distance

        # Move to the nearest city
        if nearest_city is not None:
            current_city = nearest_city
            tour.append(current_city)
            visited[current_city] = True
            total_distance += nearest_distance
        else:
            # Handle the case where there's no valid nearest city
            break

    # Complete the tour by returning to the starting city
    tour.append(0)
    total_distance += distances[current_city][0] if distances[current_city] and distances[current_city][0] is not None else 0

    return {"tour": tour, "total_distance": total_distance}

def solve_tsp_brute_force(distances):
    num_cities = len(distances)
    all_permutations = list(itertools.permutations(range(num_cities)))
    min_distance = float("inf")
    optimal_tour = None

    for permutation in all_permutations:
        current_distance = calculate_tour_distance(permutation, distances)
        if current_distance < min_distance:
            min_distance = current_distance
            optimal_tour = permutation

    return {"tour": optimal_tour, "total_distance": min_distance}

def prim_mst(graph):
    num_vertices = len(graph)
    parent = [-1] * num_vertices
    key = [float("inf")] * num_vertices
    mst_set = [False] * num_vertices

    key[0] = 0

    for count in range(num_vertices - 1):
        u = min_key(key, mst_set)
        mst_set[u] = True

        for v in range(num_vertices):
            if graph[u][v] and not mst_set[v] and graph[u][v] < key[v]:
                parent[v] = u
                key[v] = graph[u][v]

    return parent

def min_key(key, mst_set):
    num_vertices = len(key)
    min_val = float("inf")
    min_index = -1

    for v in range(num_vertices):
        if not mst_set[v] and key[v] < min_val:
            min_val = key[v]
            min_index = v

    return min_index

def dfs_on_mst(mst):
    num_vertices = len(mst)
    stack = []
    visited = [False] * num_vertices

    stack.append(0)
    visited[0] = True

    tour = []

    while stack:
        current_vertex = stack.pop()
        tour.append(current_vertex)

        for i in range(num_vertices):
            if mst[current_vertex][i] != -1 and not visited[i]:
                stack.append(i)
                visited[i] = True

    # Complete the tour by returning to the starting city
    tour.append(tour[0])

    return tour

# Your list of cities
cities = [
    {"lat": 52.5200, "lon": 13.4050},
    {"lat": 53.5511, "lon": 9.9937},
    {"lat": 48.1351, "lon": 11.5820},
    {"lat": 50.9375, "lon": 6.9603},
    {"lat": 50.1109, "lon": 8.6821},
]

# Generate the distance matrix
distances = get_distance_matrix(cities)

# Display the distance matrix
for row in distances:
    print(row)

# Solve TSP using Nearest Neighbor Algorithm
nearest_result = solve_tsp_nearest(distances)
print("Nearest Neighbor Tour:", nearest_result["tour"])
print("Nearest Neighbor Total distance:", nearest_result["total_distance"])

# Solve TSP using Brute Force Algorithm
brute_force_result = solve_tsp_brute_force(distances)
print("Brute Force Tour:", brute_force_result["tour"])
print("Brute Force Total distance:", brute_force_result["total_distance"])

# Solve TSP using Minimum Spanning Tree (MST) Algorithm
mst_result = solve_tsp_mst(distances)
print("Minimum Spanning Tree Tour:", mst_result["tour"])
print("Minimum Spanning Tree Total distance:", mst_result["total_distance"])

File: test-app/Components/3algo_backend.js
function calculateDistance(city1, city2) {
    const { lat: lat1, lon: lon1 } = city1;
    const { lat: lat2, lon: lon2 } = city2;
    const radius = 6371;  // Radius of the Earth in kilometers
    const dlat = toRadians(lat2 - lat1);
    const dlon = toRadians(lon2 - lon1);

    let a = Math.sin(dlat / 2) ** 2 + Math.cos(toRadians(lat1)) * Math.cos(toRadians(lat2)) * Math.sin(dlon / 2) ** 2;
    a = Math.min(1, a);  // Ensure a is within the valid range [0, 1]
    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    const distance = radius * c;

    return distance;
}

function toRadians(degrees) {
    return degrees * (Math.PI / 180);
}

function getDistanceMatrix(cities) {
    const numCities = cities.length;
    const distances = new Array(numCities).fill(0).map(() => new Array(numCities).fill(0));

    for (let i = 0; i < numCities; i++) {
        for (let j = i + 1; j < numCities; j++) {
            const dist = calculateDistance(cities[i], cities[j]);
            distances[i][j] = Math.round(dist);
            distances[j][i] = Math.round(dist);  // Ensure symmetry
        }
    }

    return distances;
}

function solveTSPNearest(distances) {
    const numCities = distances.length;
    const visited = new Array(numCities).fill(false);
    const tour = [];
    let totalDistance = 0;

    // Start at the first city
    let currentCity = 0;
    tour.push(currentCity);
    visited[currentCity] = true;

    // Repeat until all cities have been visited
    while (tour.length < numCities) {
        let nearestCity = null;
        let nearestDistance = Infinity;

        // Find the nearest unvisited city
        for (let city = 0; city < numCities; city++) {
            if (!visited[city] && distances[currentCity] && distances[currentCity][city] !== undefined) {
                const distance = distances[currentCity][city];
                if (distance < nearestDistance) {
                    nearestCity = city;
                    nearestDistance = distance;
                }
            }
        }

        // Move to the nearest city
        if (nearestCity !== null) {
            currentCity = nearestCity;
            tour.push(currentCity);
            visited[currentCity] = true;
            totalDistance += nearestDistance;
        } else {
            // Handle the case where there's no valid nearest city
            break;
        }
    }

    // Complete the tour by returning to the starting city
    tour.push(0);
    totalDistance += distances[currentCity] && distances[currentCity][0] !== undefined
        ? distances[currentCity][0]
        : 0;

    return { tour, totalDistance };
}

function solveTSPBruteForce(distances) {
    const numCities = distances.length;
    const allPermutations = generateAllPermutations(numCities);
    let minDistance = Infinity;
    let optimalTour;

    for (const permutation of allPermutations) {
        const currentDistance = calculateTourDistance(permutation, distances);
        if (currentDistance < minDistance) {
            minDistance = currentDistance;
            optimalTour = permutation;
        }
    }

    return { tour: optimalTour, totalDistance: minDistance };
}

function solveTSPMST(distances) {
    const numCities = distances.length;

    // Create a minimum spanning tree using Prim's algorithm
    const mst = primMST(distances);

    // Perform a Depth-First Search (DFS) on the MST to get a tour
    const tour = dfsOnMST(mst);

    // Calculate the total distance of the tour
    const totalDistance = calculateTourDistance(tour, distances);

    return { tour, totalDistance };
}

function primMST(graph) {
    const numVertices = graph.length;
    const parent = new Array(numVertices).fill(-1);
    const key = new Array(numVertices).fill(Infinity);
    const mstSet = new Array(numVertices).fill(false);

    key[0] = 0;

    for (let count = 0; count < numVertices - 1; count++) {
        const u = minKey(key, mstSet);
        mstSet[u] = true;

        for (let v = 0; v < numVertices; v++) {
            if (graph[u][v] && !mstSet[v] && graph[u][v] < key[v]) {
                parent[v] = u;
                key[v] = graph[u][v];
            }
        }
    }

    return parent;
}

function minKey(key, mstSet) {
    const numVertices = key.length;
    let min = Infinity;
    let minIndex = -1;

    for (let v = 0; v < numVertices; v++) {
        if (!mstSet[v] && key[v] < min) {
            min = key[v];
            minIndex = v;
        }
    }

    return minIndex;
}

function dfsOnMST(mst) {
    const numVertices = mst.length;
    const stack = [];
    const visited = new Array(numVertices).fill(false);

    stack.push(0);
    visited[0] = true;

    const tour = [];

    while (stack.length > 0) {
        const currentVertex = stack.pop();
        tour.push(currentVertex);

        for (let i = 0; i < numVertices; i++) {
            if (mst[currentVertex][i] !== -1 && !visited[i]) {
                stack.push(i);
                visited[i] = true;
            }
        }
    }

    // Complete the tour by returning to the starting city
    tour.push(tour[0]);

    return tour;
}

function calculateTourDistance(tour, distances) {
    let totalDistance = 0;
    for (let i = 0; i < tour.length - 1; i++) {
        const fromCity = tour[i];
        const toCity = tour[i + 1];
        totalDistance += distances[fromCity][toCity];
    }
    // Complete the tour by returning to the starting city
    totalDistance += distances[tour[tour.length - 1]][tour[0]];
    return totalDistance;
}

function generateAllPermutations(n) {
    const permutations = [];
    const indices = Array.from({ length: n }, (_, index) => index);

    function permute(arr, start, end) {
        if (start === end) {
            permutations.push([...arr]);
            return;
        }

        for (let i = start; i <= end; i++) {
            [arr[start], arr[i]] = [arr[i], arr[start]];
            permute(arr, start + 1, end);
            [arr[start], arr[i]] = [arr[i], arr[start]]; // Backtrack
        }
    }

    permute(indices, 0, n - 1);
    return permutations;
}

const cities = [
    { "lat": 52.5200, "lon": 13.4050 },
    { "lat": 53.5511, "lon": 9.9937 },
    { "lat": 48.1351, "lon": 11.5820 },
    { "lat": 50.9375, "lon": 6.9603 },
    { "lat": 50.1109, "lon": 8.6821 },
];

const distances = getDistanceMatrix(cities);

distances.forEach(row => {
    console.log(row);
});

const { tour: nearestTour, totalDistance: nearestTotalDistance } = solveTSPNearest(distances);
const { tour: bruteForceTour, totalDistance: bruteForceTotalDistance } = solveTSPBruteForce(distances);
const { tour: mstTour, totalDistance: mstTotalDistance } = solveTSPMST(distances);

console.log("Nearest Neighbor Tour:", nearestTour);
console.log("Nearest Neighbor Total distance:", nearestTotalDistance);

console.log("Brute Force Tour:", bruteForceTour);
console.log("Brute Force Total distance:", bruteForceTotalDistance);

console.log("Minimum Spanning Tree Tour:", mstTour);
console.log("Minimum Spanning Tree Total distance:", mstTotalDistance);
